File: core/pkg/runtime/variables.go
package runtime

import (
	"fmt"
	"sort"
	"strings"
)

// Level represents the scope level of a variable.
type Level int

const (
	LevelRow     Level = 1
	LevelStep    Level = 2
	LevelMission Level = 3
	LevelGlobal  Level = 4
	LevelImport  Level = 5
)

// ScopedVariables manages variables across multiple precedence levels.
type ScopedVariables struct {
	levels map[Level]map[string]string
}

// NewScopedVariables initializes a new variable scope manager.
func NewScopedVariables() *ScopedVariables {
	return &ScopedVariables{
		levels: map[Level]map[string]string{
			LevelRow:     make(map[string]string),
			LevelStep:    make(map[string]string),
			LevelMission: make(map[string]string),
			LevelGlobal:  make(map[string]string),
			LevelImport:  make(map[string]string),
		},
	}
}

// Set stores a value at the specified precedence level.
func (sv *ScopedVariables) Set(name, value string, level Level) {
	if sv.levels[level] == nil {
		sv.levels[level] = make(map[string]string)
	}
	sv.levels[level][name] = value
}

// Resolve returns the value of a variable, respecting precedence (Row > Step > Mission > Global).
func (sv *ScopedVariables) Resolve(name string) (string, bool) {
	// Priority order: 1, 2, 3, 4
	for i := Level(1); i <= 5; i++ {
		if val, ok := sv.levels[i][name]; ok {
			return val, true
		}
	}
	return "", false
}

// ResolveLevel returns the value and the level it was found at.
func (sv *ScopedVariables) ResolveLevel(name string) (string, Level, bool) {
	for i := Level(1); i <= 5; i++ {
		if val, ok := sv.levels[i][name]; ok {
			return val, i, true
		}
	}
	return "", 0, false
}

// ClearLevel wipes all variables at a specific level.
func (sv *ScopedVariables) ClearLevel(level Level) {
	sv.levels[level] = make(map[string]string)
}

// ClearAll wipes all variables across all levels.
func (sv *ScopedVariables) ClearAll() {
	for i := Level(1); i <= 5; i++ {
		sv.ClearLevel(i)
	}
}

// Flatten merges all levels into a single map, following precedence rules.
func (sv *ScopedVariables) Flatten() map[string]string {
	flat := make(map[string]string)
	// Iterate backwards (Global to Row) so higher priority overwrites.
	for i := Level(5); i >= 1; i-- {
		for k, v := range sv.levels[i] {
			flat[k] = v
		}
	}
	return flat
}

// Interpolate replaces $var, ${var}, or {var} placeholders in a string.
// It sorts keys by length (longest-first) to avoid partial matches
// (e.g., replacing $user before $user_id).
func (sv *ScopedVariables) Interpolate(s string) string {
	flat := sv.Flatten()

	// Get keys and sort by length descending
	keys := make([]string, 0, len(flat))
	for k := range flat {
		keys = append(keys, k)
	}
	sort.Slice(keys, func(i, j int) bool {
		return len(keys[i]) > len(keys[j])
	})

	for _, k := range keys {
		v := flat[k]
		// Replace braced forms first to avoid partial replacement by bare $k
		s = strings.ReplaceAll(s, "${"+k+"}", v)
		s = strings.ReplaceAll(s, "{"+k+"}", v)
		s = strings.ReplaceAll(s, "$"+k, v)
	}
	return s
}

// String returns a debug representation of all variables.
func (sv *ScopedVariables) String() string {
	var sb strings.Builder
	sb.WriteString("DEBUG VARS:\n")
	for i := Level(1); i <= 5; i++ {
		lvlName := ""
		switch i {
		case LevelRow: lvlName = "ROW"
		case LevelStep: lvlName = "STEP"
		case LevelMission: lvlName = "MISSION"
		case LevelGlobal: lvlName = "GLOBAL"
		case LevelImport: lvlName = "IMPORT"
		}
		sb.WriteString(fmt.Sprintf("  [%s]:\n", lvlName))
		for k, v := range sv.levels[i] {
			sb.WriteString(fmt.Sprintf("    %s = %s\n", k, v))
		}
	}
	return sb.String()
}

File: legacy/python/manul_engine/variables.py
# manul_engine/variables.py
"""
Strict scoped variable state management for ManulEngine.

Implements a five-level precedence hierarchy (highest → lowest):
  Level 1 — Row Vars:     Injected per-iteration from @data CSV/JSON.
  Level 2 — Step Vars:    Created mid-flight via EXTRACT or CALL PYTHON into {var}.
  Level 3 — Mission Vars: Declared in the file header via @var:.
  Level 4 — Global Vars:  Passed via CLI, env, or @before_all lifecycle hooks.
  Level 5 — Import Vars:  Declared in imported .hunt files via @var:.

resolve() strictly respects this order: a higher-level variable always
shadows a lower-level one with the same name.
"""

from __future__ import annotations


class ScopedVariables:
    """Five-level variable store with strict precedence resolution.

    Each level is a plain dict.  ``resolve(name)`` checks levels top-down;
    ``substitute(text)`` replaces all ``{var}`` placeholders using the
    resolved values.  ``set()`` writes to a specific level.
    """

    LEVEL_ROW = "row"
    LEVEL_STEP = "step"
    LEVEL_MISSION = "mission"
    LEVEL_GLOBAL = "global"
    LEVEL_IMPORT = "import"

    _LEVELS = (LEVEL_ROW, LEVEL_STEP, LEVEL_MISSION, LEVEL_GLOBAL, LEVEL_IMPORT)

    def __init__(self) -> None:
        self._stores: dict[str, dict[str, str]] = {
            self.LEVEL_ROW: {},
            self.LEVEL_STEP: {},
            self.LEVEL_MISSION: {},
            self.LEVEL_GLOBAL: {},
            self.LEVEL_IMPORT: {},
        }

    # ── Read ──────────────────────────────────────────────────────────────

    def resolve(self, name: str) -> str | None:
        """Return the value for *name* at the highest-priority level, or None."""
        for level in self._LEVELS:
            store = self._stores[level]
            if name in store:
                return store[name]
        return None

    def resolve_level(self, name: str) -> tuple[str | None, str | None]:
        """Return ``(value, level_name)`` for *name*, or ``(None, None)``."""
        for level in self._LEVELS:
            store = self._stores[level]
            if name in store:
                return store[name], level
        return None, None

    def as_flat_dict(self) -> dict[str, str]:
        """Return a merged dict respecting precedence (lowest first, highest overwrites)."""
        merged: dict[str, str] = {}
        for level in reversed(self._LEVELS):
            merged.update(self._stores[level])
        return merged

    def substitute(self, text: str) -> str:
        """Replace all ``{var}`` placeholders with resolved values."""
        flat = self.as_flat_dict()
        for k, v in flat.items():
            text = text.replace(f"{{{k}}}", str(v))
        return text

    # ── Write ─────────────────────────────────────────────────────────────

    def set(self, name: str, value: str, level: str) -> None:
        """Set a variable at a specific level."""
        if level not in self._LEVELS:
            raise ValueError(f"Unknown variable level: {level!r}")
        self._stores[level][name] = str(value)

    def set_many(self, mapping: dict[str, str], level: str) -> None:
        """Bulk-set variables at a specific level."""
        if level not in self._LEVELS:
            raise ValueError(f"Unknown variable level: {level!r}")
        self._stores[level].update({k: str(v) for k, v in mapping.items()})

    # ── Clear / reset ─────────────────────────────────────────────────────

    def clear_level(self, level: str) -> None:
        """Remove all variables at a given level."""
        if level in self._stores:
            self._stores[level].clear()

    def clear_runtime(self) -> None:
        """Clear row and step vars (used between @data iterations)."""
        self._stores[self.LEVEL_ROW].clear()
        self._stores[self.LEVEL_STEP].clear()

    def clear_all(self) -> None:
        """Clear every level."""
        for store in self._stores.values():
            store.clear()

    # ── Introspection (DEBUG VARS) ────────────────────────────────────────

    def dump(self) -> str:
        """Return a formatted multi-line string showing all levels."""
        lines: list[str] = []
        labels = {
            self.LEVEL_ROW: "Level 1 — Row Vars (@data)",
            self.LEVEL_STEP: "Level 2 — Step Vars (EXTRACT / CALL PYTHON into)",
            self.LEVEL_MISSION: "Level 3 — Mission Vars (@var:)",
            self.LEVEL_GLOBAL: "Level 4 — Global Vars (CLI / env / @before_all)",
            self.LEVEL_IMPORT: "Level 5 — Import Vars (@import file @var:)",
        }
        for level in self._LEVELS:
            store = self._stores[level]
            label = labels[level]
            lines.append(f"  ┌─ {label}")
            if store:
                for k, v in store.items():
                    lines.append(f"  │  {{{k}}} = {v}")
            else:
                lines.append("  │  (empty)")
            lines.append(f"  └{'─' * 50}")
        return "\n".join(lines)

    # ── dict-like compatibility layer ─────────────────────────────────────
    # ManulEngine historically uses self.memory as a plain dict.
    # These methods let ScopedVariables be used in place of self.memory
    # while existing code that reads via dict[key] or .get() still works.

    def __contains__(self, name: str) -> bool:
        return self.resolve(name) is not None

    def __getitem__(self, name: str) -> str:
        val = self.resolve(name)
        if val is None:
            raise KeyError(name)
        return val

    def __setitem__(self, name: str, value: str) -> None:
        # Default write target: step level (runtime assignments)
        self._stores[self.LEVEL_STEP][name] = str(value)

    def get(self, name: str, default: str | None = None) -> str | None:
        val = self.resolve(name)
        return val if val is not None else default

    def items(self):
        return self.as_flat_dict().items()

    def keys(self):
        return self.as_flat_dict().keys()

    def values(self):
        return self.as_flat_dict().values()

    def update(self, mapping) -> None:
        # Default update goes to step level for backward compat.
        if isinstance(mapping, dict):
            self._stores[self.LEVEL_STEP].update({k: str(v) for k, v in mapping.items()})

    def clear(self) -> None:
        self.clear_all()

    def __eq__(self, other) -> bool:
        if isinstance(other, ScopedVariables):
            return self._stores == other._stores
        if isinstance(other, dict):
            return self.as_flat_dict() == other
        return NotImplemented

    def __repr__(self) -> str:
        return f"ScopedVariables({self.as_flat_dict()!r})"
